Sources/SourceKitLSP/Swift/SwiftLanguageServer.swift

```diff
@@ -499,20 +419,75 @@ extension SwiftLanguageServer {
     req[keys.request] = self.requests.editor_close
     req[keys.name] = uri.pseudoPath
 
-    // Clear settings that should not be cached for closed documents.
-    self.currentDiagnostics[uri] = nil
-
     _ = try? self.sourcekitd.sendSync(req)
 
     await semanticTokensManager.discardSemanticTokens(for: note.textDocument.uri)
   }
 
+  /// Cancels any in-flight tasks to send a `PublishedDiagnosticsNotification` after edits.
+  ///
+  /// - Important: Cancelling in-flight publish diagnostics tasks is required before any edit. See the comment inside
+  ///   `documentDiagnostic` for more details.
+  private func cancelInFlightPublishDiagnosticsTask(for document: DocumentURI) {
+    if let inFlightTask = inFlightPublishDiagnosticsTasks[document] {
+      inFlightTask.cancel()
+    }
+  }
+
+  /// If the client doesn't support pull diagnostics, compute diagnostics for the latest version of the given document
+  /// and send a `PublishDiagnosticsNotification` to the client for it.
+  private func publishDiagnosticsIfNeeded(for document: DocumentURI) {
+    guard enablePublishDiagnostics else {
+      return
+    }
+    guard diagnosticsEnabled(for: document) else {
+      log("Ignoring diagnostics for blacklisted file \(document.pseudoPath)", level: .debug)
+      return
+    }
+    cancelInFlightPublishDiagnosticsTask(for: document)
+    inFlightPublishDiagnosticsTasks[document] = Task(priority: .medium) { [weak self] in
+      guard let self, let sourceKitServer = await self.sourceKitServer else {
+        log("Cannot produce PublishDiagnosticsNotification because sourceKitServer got deallocated", level: .error)
+        return
+      }
+      do {
+        // Sleep for a little bit until triggering the diagnostic generation. This effectively de-bounces diagnostic
+        // generation since any later edit will cancel the previous in-flight task, which will thus never go on to send
+        // the `DocumentDiagnosticsRequest`.
+        try await Task.sleep(
+          nanoseconds: UInt64(sourceKitServer.options.swiftPublishDiagnosticsDebounceDuration * 1_000_000_000)
+        )
+      } catch {
+        return
+      }
+      do {
+        // Run the diagnostic report request through `SourceKitServer` so that it gets handled on `SourceKitServer`'s
+        // `messageHandlingQueue`. This ensures that the diagnostics request will not run concurrently with any document
+        // edits.
+        let diagnosticsResult = try await self.documentDiagnostic(DocumentDiagnosticsRequest(textDocument: TextDocumentIdentifier(document)))
+        guard case .full(let diagnosticReport) = diagnosticsResult else {
+          log("Expected a full diagnostic report be returned by the diagnostics request", level: .error)
```

Comment:
I changed the phrasing to

```swift
logger.fault("Expected that `documentDiagnostic` always returns a full diagnostic report")
```

Reply:
IMO it would be nicer to avoid this case altogether by defining e.g:

 ```swift
func documentDiagnostics(for textDocument: TextDocumentIdentifier) async throws -> RelatedFullDocumentDiagnosticReport
```

and then have `documentDiagnostic` forward onto it for requests, but have this logic call the version that always returns a full report.

Reply:
Very good idea 👍🏽

Reply:
So the former, thanks 👍